source/common/http/utility.cc

```diff
@@ -380,6 +381,128 @@ Utility::getLastAddressFromXFF(const Http::HeaderMap& request_headers, uint32_t
+      absl::StrSplit(connection_header_value.getStringView(), ',');
+
+  // If we have 10 or more nominated headers, fail this request
+  if (connection_header_tokens.size() >= 10) {
+    ENVOY_LOG_MISC(trace, "Too many nominated headers in request");
+    return false;
+  }
+
+  // Split the connection header and evaluate each nominated header
+  for (const auto& token : connection_header_tokens) {
+
+    const absl::string_view token_sv = StringUtil::trim(token);
+
+    // If the Connection token value is not a nominated header, ignore it here since
+    // the connection header is removed elsewhere when the H1 request is upgraded to H2
+    if (StringUtil::CaseInsensitiveCompare()(token_sv, cv.Close) ||
+        StringUtil::CaseInsensitiveCompare()(token_sv, cv.Http2Settings) ||
+        StringUtil::CaseInsensitiveCompare()(token_sv, cv.KeepAlive) ||
+        StringUtil::CaseInsensitiveCompare()(token_sv, cv.Upgrade)) {
+      continue;
+    }
+
+    // Build the LowerCaseString for header lookup
+    const std::string header_to_remove = std::string(token_sv);
+    const LowerCaseString lcs_header_to_remove(header_to_remove);
+
+    // By default we will remove any nominated headers
+    bool keep_header = false;
+    StringUtil::CaseUnorderedSet tokens_to_remove{};
+
+    // Determine whether the nominated header contains invalid values
+    HeaderEntry* nominated_header = NULL;
+
+    if (StringUtil::CaseInsensitiveCompare()(token_sv, Http::Headers::get().Connection.get())) {
+      // Remove the connection header from the nominated tokens if it's self nominated
+      // The connection header itself is *not removed*
+      ENVOY_LOG_MISC(trace, "Skipping self nominated header [{}]", token_sv);
+      keep_header = true;
+      headers_to_remove.emplace(token_sv);
+
+    } else if (StringUtil::CaseInsensitiveCompare()(token_sv,
+                                                    Http::Headers::get().Connection.get()) ||
+               StringUtil::CaseInsensitiveCompare()(token_sv,
+                                                    Http::Headers::get().ForwardedFor.get()) ||
+               StringUtil::CaseInsensitiveCompare()(token_sv,
+                                                    Http::Headers::get().ForwardedHost.get()) ||
+               StringUtil::CaseInsensitiveCompare()(token_sv,
+                                                    Http::Headers::get().ForwardedProto.get()) ||
+               !token_sv.find(':')) {
+      // If pseudo headers or X-Forwarded* headers are nominated, this could be
+      // an invalid request. Reject the request
+      return false;
+
+    } else {
+      // Examine the value of all other nominated headers
+      nominated_header = headers.get(lcs_header_to_remove);
+    }
+
+    if (nominated_header) {
+      const HeaderString& nominated_header_value = nominated_header->value();
+      for (const auto& header_value : absl::StrSplit(nominated_header_value.getStringView(), ',')) {
```

Comment:
I'm a bit concerned that it may DoS vector by creating long TE with a lot of tokens and adding Connection: TE header. Perhaps an additional check that TE is less than N (i.e. 256) bytes long would be appropriate as TE normally should not be long? Or replace full blown split with something lighter that just checks if the "trailers" keyword is present.

Reply:
I will check the length.   If 'trailers' is present, we'd still have to split and sanitize.